fastparquet/json.py

```diff
@@ -0,0 +1,97 @@
+import logging
+from abc import ABC, abstractmethod
+from functools import lru_cache
+
+logger = logging.getLogger("parquet")
+
+
+class BaseImpl(ABC):
+    @abstractmethod
+    def dumps(self, data):
+        """Serialize ``obj`` to a JSON formatted bytes instance containing UTF-8 data."""
+
+    @abstractmethod
+    def loads(self, s):
+        """Deserialize ``s`` (str, bytes or bytearray containing JSON) to a Python object."""
+
+
+class OrjsonImpl(BaseImpl):
+    def __init__(self):
+        import orjson
+
+        logger.debug("Using orjson encoder/decoder")
+        self.api = orjson
+
+    def dumps(self, data):
+        return self.api.dumps(data, option=self.api.OPT_SERIALIZE_NUMPY)
+
+    def loads(self, s):
+        return self.api.loads(s)
+
+
+class UjsonImpl(BaseImpl):
+    def __init__(self):
+        import ujson
+
+        logger.debug("Using ujson encoder/decoder")
+        self.api = ujson
+
+    def dumps(self, data):
+        return self.api.dumps(
+            data,
+            ensure_ascii=False,
+            escape_forward_slashes=False,
+        ).encode("utf-8")
+
+    def loads(self, s):
+        return self.api.loads(s)
+
+
+class RapidjsonImpl(BaseImpl):
+    def __init__(self):
+        import rapidjson
+
+        logger.debug("Using rapidjson encoder/decoder")
+        self.api = rapidjson
+
+    def dumps(self, data):
+        return self.api.dumps(data, ensure_ascii=False).encode("utf-8")
+
+    def loads(self, s):
+        return self.api.loads(s)
+
+
+class JsonImpl(BaseImpl):
+    def __init__(self):
+        import json
+
+        logger.debug("Using json encoder/decoder")
+        self.api = json
+
+    def dumps(self, data):
+        return self.api.dumps(data, separators=(",", ":")).encode("utf-8")
+
+    def loads(self, s):
+        return self.api.loads(s)
+
+
+@lru_cache(maxsize=None)
+def _get_json_impl():
+    """Return the first available json encoder/decoder implementation."""
+    for engine_class in [OrjsonImpl, UjsonImpl, RapidjsonImpl]:
```

Comment:
Changing a global variable wouldn't play nicely when spawning subprocesses with multiprocessing, so I tried to implement the preference with an env variable named `FASTPARQUET_JSON_CODEC` that should be propagated to the subprocesses at the moment of the spawn or fork.

A similar env variable `FASTPARQUET_DATAPAGE_V2` is already used for other purposes, so this approach seems consistent. Any thoughts?

I kept the same fallback logic in case the env variable is empty or undefined.

I also added a `clear_cached_codec()` function that wraps `.cache_clear()` and it could be used to unset the current implementation, but probably it can be useful mainly for debugging and playing with the different libraries.

Reply:
I did the implementation mentioned in the previous edited comment and removed `clear_cached_codec()`, now it's possible to dynamically change the codec by setting a different value for the env variable:
```
os.environ["FASTPARQUET_JSON_CODEC"] = "ujson"
... <test code>
os.environ["FASTPARQUET_JSON_CODEC"] = "rapidjson"
... <test code>
```